polyroots: Exclude zero roots of resultant in the Descartes method for quartics

normalhuman opened the pull request:
#### References to other Issues or PRs

Fixes #14517

#### Brief description of what is fixed or changed

The method `_roots_quartic_euler`  of polyroots picks the largest rational root of the resolvent `64*x**3 + 32*p*x**2 + (4*p**2 - 16*r)*x - q**2` expecting it to be positive because the constant term is `-q**2`. This expectation is unwarranted since the coefficients may be complex, or q could be zero. This is not necessarily a problem, unless the root is zero. Zero root of resolvent cannot be used by this method, as it's being divided by.  

Now the zero root is filtered out. Taking the maximum is fine after that; positive roots are preferred.

normalhuman committed:
Exclude zero roots of resultant in the Descartes method for quartics

The method `_roots_quartic_euler` picks the largest rational root of the
resolvent `64*x**3 + 32*p*x**2 + (4*p**2 - 16*r)*x - q**2` expecting it
to be positive because the constant term is `-q**2`. This expectation is
unwarranted since the coefficients may be complex, or q could be zero.

Instead, the zero root(s) should be explicitly filtered out. Taking
maximum is fine after that.

jksuom commented:
There will be a zero root only if `q == 0`. In this case, the depressed quartic equation (with no `x**3` term) has no linear term and becomes a quadratic equation in `z**2`. Classically, this is treated as a special case.

normalhuman committed:
Use expand before comparing the resolvent coefficients to 0

normalhuman commented:
Right, so the issue was higher up, in `roots_quartic`. After the resolvent coefficients were computed, 
```
        f = c + a*(a2/8 - b/2)
        g = d - a*(a*(3*a2/256 - b/16) + c/4)
```
they were zero-tested with f is S.Zero and g is S.Zero. In the above example, f is 
```
-2000 - 800*I + (-10 - 4*I)*(-S(421)/2 - 10*I + (-10 - 4*I)**2/8)
```
which isn't obviously 0. I added `expand` to this step, and added another reducible-to-biquadratic test case for polyroots.  

But kept the filter for zero resolvent roots in the helper routine, because whatever zero checks are done in `roots_quartic`, there is a possibility they will fail, and the resolvent has a zero root.

smichr commented:
> I added expand to this step

`_mexpand(expr, recursive=True)` should be sufficient instead of using the wide spectrum `expand`.
```python
>>> source(_mexpand)
In file: \sympy\core\function.py
def _mexpand(expr, recursive=False):
    # expand multinomials and then expand products; this may not always
    # be sufficient to give a fully expanded expression (see
    # test_issue_8247_8354 in test_arit)
    if expr is None:
        return
    was = None
    while was != expr:
        was, expr = expr, expand_mul(expand_multinomial(expr))
        if not recursive:
            break
    return expr
```

normalhuman commented:
It is sufficient, but I wonder why we are importing a private function to use in other modules of SymPy. Should it be renamed to `expand_mul_recursive` if it's widely useful?

smichr commented:
Or even `expand_poly`?

smichr commented:
> importing a private function

It was created as a private function at a time when its general utility was not known well enough to create a new method. That shouldn't stop you from using it if it is what you need for this code. Or else feel free to add a new method as we have suggested (though I prefer a shorter name).

normalhuman committed:
use _mexpand instead of expand

normalhuman commented:
I replaced expand with `_mexpand`. 

Left the renaming to `expand_poly` for another time, because one of the files involved is ode.py, and ongoing work there has enough merge conflicts as is.
